### PR TITLE
Optimize cancellation handling in Hosting

### PR DESCRIPTION
**Pull request by pentp**
Avoid creating unnecessary `CancellationTokenSource` links and use `ConfigureAwaitOptions.SuppressThrowing` in a few places.

### CODE REVIEW

**Comment by steveharter** [comment]
Was this PR created because of something you noticed and measured and\or does it have some background to it not captured here?

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs [hunk 1]
```diff
@@ -377,7 +360,14 @@ public async Task StopAsync(CancellationToken cancellationToken = default)
                     }
                     catch (Exception ex)
                     {
-                        exceptions.AddRange(groupedTasks.Exception?.InnerExceptions ?? new[] { ex }.AsEnumerable());
+                        if (groupedTasks.IsFaulted)
```
Comment (steveharter) [inline src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:363]: Why was this changed?

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs [hunk 2]
```diff
@@ -363,7 +346,7 @@ public async Task StopAsync(CancellationToken cancellationToken = default)
                     {
                         // The task encountered an await; add it to a list to run concurrently.
                         tasks ??= new();
-                        tasks.Add(Task.Run(() => task, token));
+                        tasks.Add(task);
```
Comment (steveharter) [inline src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:349]: Reference: https://github.com/dotnet/runtime/issues/93970

Comment (pentp) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:349]: This `Task.Run` is basically a no-op - the `token` applies only to the running of the lambda, so there's theoretically a way for the cancellation to apply, but it's close to zero and doesn't do what the original author probably thought (it doesn't apply to the returned task).

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs (hunk 1 above)
Comment (pentp) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:363]: To reduce allocations.

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs [hunk 3]
```diff
@@ -421,34 +411,30 @@ public async Task StopAsync(CancellationToken cancellationToken = default)
...
 
-            static async ValueTask DisposeAsync(object o)
+            static ValueTask DisposeAsync(object o)
             {
                 switch (o)
                 {
                     case IAsyncDisposable asyncDisposable:
-                        await asyncDisposable.DisposeAsync().ConfigureAwait(false);
-                        break;
+                        return asyncDisposable.DisposeAsync();
```
[the remark this answers is not in the document]
Comment (pentp) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:432]: There's no semantic change here, `DisposeAsync` callers continue to `await` with `ConfigureAwait(false)`.

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs (hunk 1 above)
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:363]: Sure; I was thinking of the new LOC for checking `IsFaulted` - when will it be `false`?

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:432]: There is an unnecessary `await` + `ConfigureAwait(false)` for the `IAsyncDisposable` case but I think that may have been to prevent awaiting a default task for the `IDisposable` case?

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/ApplicationLifetime.cs [hunk 4]
```diff
@@ -107,17 +107,5 @@ public void NotifyStopped()
                                          ex);
             }
         }
-
-        private static void ExecuteHandlers(CancellationTokenSource cancel)
-        {
-            // Noop if this is already cancelled
-            if (cancel.IsCancellationRequested)
```
Comment (steveharter) [inline src/libraries/Microsoft.Extensions.Hosting/src/Internal/ApplicationLifetime.cs:114]: What is the reason to no longer include this "noop" case?

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs (hunk 2 above)
Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:349]: It believe it is a no-op now. I believe it used to support running the sync portion of each task (before the first await) in parallel before.

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (pentp) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:432]: `default` for `ValueTask` is the same as `Task.CompletedTask` for `Task`

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/ApplicationLifetime.cs (hunk 4 above)
Comment (pentp) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/ApplicationLifetime.cs:114]: `CancellationTokenSource` has the exact same check already, so calling it multiple times has no effect.

Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/ApplicationLifetime.cs:114]: OK, I verified this is correct for both Core and .NET Framework.

**Review by steveharter — APPROVED** [review]
LGTM; thanks
